tck/src/main/java/org/eclipse/microprofile/config/tck/converters/ConvertedNullValueTest.java

```diff
@@ -0,0 +1,95 @@
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.eclipse.microprofile.config.tck.converter;
+
+import java.util.NoSuchElementException;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.eclipse.microprofile.config.spi.Converter;
+import org.eclipse.microprofile.config.tck.converters.Pizza;
+import org.eclipse.microprofile.config.tck.converters.PizzaConverter;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.testng.Arquillian;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+/**
+ * This is to test that when a converter returns null for some config value, null will be assigned. This means the
+ * property still exists but the value is set to null on purpose.
+ */
+public class ConvertedNullValueTest extends Arquillian {
+    private @Inject Config config;
+    private @Inject MyBean myBean;
+
+    @Deployment
+    public static Archive deployment() {
+        return ShrinkWrap
+                .create(WebArchive.class, "ConvertedNullValueTest.war")
+                .addClasses(ConvertedNullValueTest.class, Pizza.class, PizzaConverter.class, MyBean.class)
+                .addAsResource(
+                        new StringAsset(
+                                "partial.pizza=cheese"),
+                        "META-INF/microprofile-config.properties")
+                .addAsServiceProvider(Converter.class, PizzaConverter.class)
+                .addAsWebInfResource(EmptyAsset.INSTANCE, "beans.xml");
+
+    }
+
+    @Test
+    public void testDefaultValue() {
+        // The converter returns null as the converter returns null if the property does not contain :
+        // Therefore, the defaultValue was used.
+        Assert.assertEquals(myBean.getPizza(), new Pizza("chicken", "medium"));
+    }
+
+    @Test
+    public void testGetValue() {
+        // The converter returns null as the converter returns null if the property does not contain :
+        // Therefore, it will be treated as non-existence.
+        Assert.assertThrows(NoSuchElementException.class, () -> config.getValue("partial.pizza", Pizza.class));
```

Comment:
While this should work when calling `Config.getValue`, we use the same property name in the `private @Inject @ConfigProperty(name = "partial.pizza", defaultValue = "medium:chicken") Pizza myPizza;` injection point.

In this case, the resolved property is null, so we should get a `DeploymentException`, right?

Reply:
`DeploymentException` is for CDI usage while the programmatic apis should work without CDI. It does not make sense to use the CDI `DeploymentException`. The javadoc of `getValue` states:
```
    /**
     * Return the resolved property value with the specified type for the specified property name from the underlying
     * {@linkplain ConfigSource configuration sources}.
     * <p>
     * The configuration value is not guaranteed to be cached by the implementation, and may be expensive to compute;
     * therefore, if the returned value is intended to be frequently used, callers should consider storing rather than
     * recomputing it.
     * <p>
     * The result of this method is identical to the result of calling
     * {@code getOptionalValue(propertyName, propertyType).get()}. In particular, If the given property name or the
     * value element of this property does not exist, the {@link java.util.NoSuchElementException} is thrown. This
     * method never returns {@code null}.
     *
     * @param <T>
     *            The property type
     * @param propertyName
     *            The configuration property name
     * @param propertyType
     *            The type into which the resolved property value should get converted
     * @return the resolved property value as an instance of the requested type (not {@code null})
     * @throws IllegalArgumentException
     * 
     * @throws java.util.NoSuchElementException
     *             if the property is not defined or is defined as an empty string or the converter returns {@code null}
     */
    <T> T getValue(String propertyName, Class<T> propertyType);

Reply:
Besides DeploymentException means your microservice fail to start. In the programmatic apis, the exception is only thrown when you call the method.

Reply:
It will not, because the defaultValue will be used. Therefore, you will not get `DeploymentException`.

Reply:
Sorry, I didn't explain myself correctly. In this case, the `partial.pizza` will resolve to `cheese` from `microprofile-config.properties`, on the injection point:

```java
private @Inject @ConfigProperty(name = "partial.pizza", defaultValue = "medium:chicken") Pizza myPizza;
```

This will fail the conversion and return `null`, so we fall into the rules detailed in https://github.com/eclipse/microprofile-config/issues/446, resolving to an empty value which should throw a `DeploymentException` when validating the injection point in `MyBean`.

Reply:
I see you are confused. For `DeploymentException`, it is only applicable to required config properties. Once you have a nonnull `defaultValue` specified on the injection point, the property is optional as you can always fallback to the `defaultValue`, which is more like an optional with a default value supplied. Therefore, you will not get `DeploymentException`. The `defaultValue` will be used. This whole PR is to further clarify that the null value resulted from converter will behave as the property does not exist. Please re-read the issue #608 . There was two options: option a and b. I created #611 and this. Since you and @jbee are not happy with #611, hence this PR.

Reply:
Sorry, once again, I think I was not clear.

In the test definition you set `partial.pizza=cheese`: https://github.com/eclipse/microprofile-config/pull/618/files#r504057435

Meaning that the config lookup will resolve to `cheese` and not to the defaultValue. When `cheese` is used in the Converter, it returns null, so we get the `DeploymentException`.

I believe we need to remove the `microprofile-config.properties` values in the test setup.

Reply:
@radcortez the point is that whether null returned from converter make the defaultValue to be used. As an end user, what behaviour do you want to see?
From the injection point of view, do you want the defaultValue to be used or you want to get DeploymentException? For programmatic lookup, when a customer does `config.getOptionValue().orElse()` and provide the default, will the default to be used?

The bottom line is that:
Do we want to treat this use case as if the config value does not exist?

Reply:
I think it needs to be consistent in the end. I look into `defaultValue` as just any other source, which is the last value we take if we don't find anything in previous sources, and only then apply conversion.

Now imagine we have the following scenario:

```
EnvSource - `partial.pizza=cheese`
microprofile-config.properties - `partial.pizza=medium,chicken`
default - `partial.pizza=big,cheese `
```

The EnvSource value will be taken first, but the value cannot be converted, do we keep falling back and read and try to convert from `microprofile-config.properties`? Or do we fallback to the default?